Dtype check not working for assert frame (not) equal

### Checks

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
from polars.testing import assert_frame_not_equal

schema_a = {
    "int_a": pl.Int16,
    "int_b": pl.Int16,
    "int_c": pl.Int64,
}

schema_b = {
    "int_a": pl.Int64,
    "int_b": pl.Int16,
    "int_c": pl.Int64,
}

a = pl.DataFrame({}, schema=schema_a)
b = pl.DataFrame({}, schema=schema_b)

assert_frame_not_equal(
    left=a,
    right=b,
    check_column_order=False,
    check_dtypes=True,
)
```


### Log output

```shell
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[11], line 18
     15 a = pl.DataFrame({}, schema=schema_a)
     16 b = pl.DataFrame({}, schema=schema_b)
---> 18 assert_frame_not_equal(
     19     left=a,
     20     right=b,
     21     check_column_order=False,
     22     check_dtypes=True,
     23 )

File ~/Workspace/SPF/spf_data_provider/.venv/lib/python3.11/site-packages/polars/_utils/deprecation.py:128, in deprecate_renamed_parameter.<locals>.decorate.<locals>.wrapper(*args, **kwargs)
    123 @wraps(function)
    124 def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
    125     _rename_keyword_argument(
    126         old_name, new_name, kwargs, function.__qualname__, version
    127     )
--> 128     return function(*args, **kwargs)

File ~/Workspace/SPF/spf_data_provider/.venv/lib/python3.11/site-packages/polars/_utils/deprecation.py:128, in deprecate_renamed_parameter.<locals>.decorate.<locals>.wrapper(*args, **kwargs)
    123 @wraps(function)
    124 def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
    125     _rename_keyword_argument(
    126         old_name, new_name, kwargs, function.__qualname__, version
    127     )
--> 128     return function(*args, **kwargs)

File ~/Workspace/SPF/spf_data_provider/.venv/lib/python3.11/site-packages/polars/_utils/deprecation.py:128, in deprecate_renamed_parameter.<locals>.decorate.<locals>.wrapper(*args, **kwargs)
    123 @wraps(function)
    124 def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
    125     _rename_keyword_argument(
    126         old_name, new_name, kwargs, function.__qualname__, version
    127     )
--> 128     return function(*args, **kwargs)

File ~/Workspace/SPF/spf_data_provider/.venv/lib/python3.11/site-packages/polars/testing/asserts/frame.py:231, in assert_frame_not_equal(left, right, check_row_order, check_column_order, check_dtypes, check_exact, rel_tol, abs_tol, categorical_as_str)
    229 objects = "LazyFrames" if lazy else "DataFrames"
    230 msg = f"{objects} are equal (but are expected not to be)"
--> 231 raise AssertionError(msg)

AssertionError: DataFrames are equal (but are expected not to be)
```

### Issue description

In polars.testing _assert_frame_equal_ and _assert_frame_not_equal_ currently do not correctly check the dtypes if the column order does not matter. I noticed that for the Int dytes comparing Int16 with Int64.
If the param _check_column_order_ is set to True everything works correctly so it is most probably connected to that.

### Expected behavior

assert_frame_equal: an AssertionError is only thrown if the dtypes are different
assert_frame_not_equal: an AssertionError is thrown if any dtype is differnt

### Installed versions

<details>

```
--------Version info---------
Polars:              1.38.1
Index type:          UInt32
Platform:            macOS-15.7.3-arm64-arm-64bit
Python:              3.11.9 (main, Aug 14 2024, 04:17:21) [Clang 18.1.8 ]
Runtime:             rt32

----Optional dependencies----
Azure CLI            <not installed>
adbc_driver_manager  <not installed>
altair               <not installed>
azure.identity       1.25.1
boto3                <not installed>
cloudpickle          <not installed>
connectorx           <not installed>
deltalake            <not installed>
fastexcel            0.19.0
fsspec               2026.1.0
gevent               <not installed>
google.auth          <not installed>
great_tables         <not installed>
matplotlib           <not installed>
numpy                1.26.4
openpyxl             <not installed>
pandas               3.0.0
polars_cloud         <not installed>
pyarrow              20.0.0
pydantic             <not installed>
pyiceberg            <not installed>
sqlalchemy           <not installed>
torch                <not installed>
xlsx2csv             0.8.6
xlsxwriter           3.2.9
```

</details>

--- a/crates/polars-testing/src/asserts/utils.rs
+++ b/crates/polars-testing/src/asserts/utils.rs
@@ -767,19 +767,17 @@ fn assert_dataframe_schema_equal(
                     format!("{:?}", right_dtypes_ordered)
                 ));
             }
-        } else {
-            if left_schema.len() != right_schema.len()
-                || left_schema
-                    .iter()
-                    .any(|(name, dtype)| right_schema.get(name) != Some(dtype))
-            {
-                return Err(polars_err!(
-                    assertion_error = "DataFrames",
-                    "dtypes do not match",
-                    format!("{:?}", left_schema),
-                    format!("{:?}", right_schema)
-                ));
-            }
+        } else if left_schema.len() != right_schema.len()
+            || left_schema
+                .iter()
+                .any(|(name, dtype)| right_schema.get(name) != Some(dtype))
+        {
+            return Err(polars_err!(
+                assertion_error = "DataFrames",
+                "dtypes do not match",
+                format!("{:?}", left_schema),
+                format!("{:?}", right_schema)
+            ));
         }
     }
 

--- a/crates/polars-testing/src/asserts/utils.rs
+++ b/crates/polars-testing/src/asserts/utils.rs
@@ -768,16 +768,16 @@ fn assert_dataframe_schema_equal(
                 ));
             }
         } else {
-            let left_dtypes_unordered: PlHashMap<&PlSmallStr, &DataType> =
-                left_schema.iter().collect();
-            let right_dtypes_unordered: PlHashMap<&PlSmallStr, &DataType> =
-                right_schema.iter().collect();
-            if left_dtypes_unordered != right_dtypes_unordered {
+            if left_schema.len() != right_schema.len()
+                || left_schema
+                    .iter()
+                    .any(|(name, dtype)| right_schema.get(name) != Some(dtype))
+            {
                 return Err(polars_err!(
                     assertion_error = "DataFrames",
                     "dtypes do not match",
-                    format!("{:?}", left_dtypes_unordered),
-                    format!("{:?}", right_dtypes_unordered)
+                    format!("{:?}", left_schema),
+                    format!("{:?}", right_schema)
                 ));
             }
         }

--- a/crates/polars-testing/src/asserts/utils.rs
+++ b/crates/polars-testing/src/asserts/utils.rs
@@ -768,14 +768,16 @@ fn assert_dataframe_schema_equal(
                 ));
             }
         } else {
-            let left_dtypes: PlHashSet<DataType> = left.dtypes().into_iter().collect();
-            let right_dtypes: PlHashSet<DataType> = right.dtypes().into_iter().collect();
-            if left_dtypes != right_dtypes {
+            let left_dtypes_unordered: PlHashMap<&PlSmallStr, &DataType> =
+                left_schema.iter().collect();
+            let right_dtypes_unordered: PlHashMap<&PlSmallStr, &DataType> =
+                right_schema.iter().collect();
+            if left_dtypes_unordered != right_dtypes_unordered {
                 return Err(polars_err!(
                     assertion_error = "DataFrames",
                     "dtypes do not match",
-                    format!("{:?}", left_dtypes),
-                    format!("{:?}", right_dtypes)
+                    format!("{:?}", left_dtypes_unordered),
+                    format!("{:?}", right_dtypes_unordered)
                 ));
             }
         }

--- a/py-polars/tests/unit/testing/test_assert_frame_equal.py
+++ b/py-polars/tests/unit/testing/test_assert_frame_equal.py
@@ -478,3 +478,37 @@ def test_frame_schema_fail():
     assert "AssertionError: DataFrames are equal" in stdout
     assert "AssertionError: inputs are different (unexpected input types)" in stdout
     assert "AssertionError: DataFrames are different (dtypes do not match)" in stdout
+
+
+def test_dtype_check_for_assert_frame_26507() -> None:
+    schema_a = {
+        "int_a": pl.Int16,
+        "int_b": pl.Int16,
+        "int_c": pl.Int64,
+    }
+
+    schema_b = {
+        "int_a": pl.Int64,
+        "int_b": pl.Int16,
+        "int_c": pl.Int64,
+    }
+
+    a = pl.DataFrame({}, schema=schema_a)
+    b = pl.DataFrame({}, schema=schema_b)
+
+    # SHOULD NOT raise an error since frames have different dtypes
+    assert_frame_not_equal(
+        left=a,
+        right=b,
+        check_column_order=False,
+        check_dtypes=True,
+    )
+
+    # SHOULD raise an error since frames have different dtypes
+    with pytest.raises(AssertionError):
+        assert_frame_equal(
+            left=a,
+            right=b,
+            check_column_order=False,
+            check_dtypes=True,
+        )

